runtime/queries/column_timeseries.go

```diff
@@ -0,0 +1,537 @@
+	prefix string,
+	dimensionJoiner string,
+) string {
+	if len(dimensionValues) == 0 {
+		return ""
+	}
+	var result string
+	conditions := make([]string, 3)
+	if len(dimensionValues) > 0 {
+		result += " ( "
+	}
+	for i, dv := range dimensionValues {
+		escapedName := EscapeSingleQuotes(dv.Name)
+		var nulls bool
+		var notNulls bool
+		for _, iv := range dv.In {
+			if _, ok := iv.Kind.(*structpb.Value_NullValue); ok {
+				nulls = true
+			} else {
+				notNulls = true
+			}
+		}
+		conditions = conditions[:0]
+		if notNulls {
+			var inClause = escapedName + " " + prefix + " IN ("
+			for j, iv := range dv.In {
+				if _, ok := iv.Kind.(*structpb.Value_NullValue); !ok {
+					inClause += "'" + EscapeSingleQuotes(iv.GetStringValue()) + "'"
+					if j < len(dv.In)-1 {
+						inClause += ", "
+					}
+				}
+			}
+			inClause += ")"
+			conditions = append(conditions, inClause)
+		}
+		if nulls {
+			var nullClause = escapedName + " IS " + prefix + " NULL"
+			conditions = append(conditions, nullClause)
+		}
+		if len(dv.Like) > 0 {
+			var likeClause string
+			for j, lv := range dv.Like {
+				if lv.GetKind() == nil {
+					continue
+				}
+				likeClause += escapedName + " " + prefix + " ILIKE '" + EscapeSingleQuotes(lv.GetStringValue()) + "'"
+				if j < len(dv.Like)-1 {
+					likeClause += " OR "
+				}
+			}
+			conditions = append(conditions, likeClause)
+		}
+		result += strings.Join(conditions, " "+dimensionJoiner+" ")
+		if i < len(dimensionValues)-1 {
+			result += ") AND ("
+		}
+	}
+	result += " ) "
+
+	return result
+}
+
+func getFilterFromMetricsViewFilters(filters *runtimev1.MetricsViewRequestFilter) string {
+	includeFilters := getFilterFromDimensionValuesFilter(filters.Include, "", "OR")
+	excludeFilters := getFilterFromDimensionValuesFilter(filters.Exclude, "NOT", "AND")
+	if includeFilters != "" && excludeFilters != "" {
+		return " ( " + includeFilters + ") AND (" + excludeFilters + ")"
+	} else if includeFilters != "" {
+		return includeFilters
+	} else if excludeFilters != "" {
+		return excludeFilters
+	} else {
+		return ""
+	}
+}
+
+func getFallbackMeasureName(index int, sqlName string) string {
+	if sqlName == "" {
+		s := fmt.Sprintf("measure_%d", index)
+		return s
+	} else {
+		return sqlName
+	}
+}
+
+var countName = "count"
+
+func normaliseMeasures(measures []*runtimev1.GenerateTimeSeriesRequest_BasicMeasure) []*runtimev1.GenerateTimeSeriesRequest_BasicMeasure {
+	if len(measures) == 0 {
+		return []*runtimev1.GenerateTimeSeriesRequest_BasicMeasure{
+			{
+				Expression: "count(*)",
+				SqlName:    countName,
+				Id:         "",
+			},
+		}
+	}
+	for i, measure := range measures {
+		measure.SqlName = getFallbackMeasureName(i, measure.SqlName)
+	}
+	return measures
+}
+
+const IsoFormat string = "2006-01-02T15:04:05.000Z"
+
+func sMap(k string, v float64) map[string]float64 {
+	m := make(map[string]float64, 1)
+	m[k] = v
+	return m
+}
+
+func convertToDateTruncSpecifier(specifier runtimev1.TimeGrain) string {
```

Comment:
I wonder if we need similar method in other APIs as well, If yes, we can move  to timeutil

Reply:
Looks like it's only used for this particular query.

But for shared SQL-building util functions (i.e. that are used by multiple queries), they should be in a central place. Maybe in `queries/sqlutil.go`?